### ISSUE TITLE
508-defect-2 [COGNITION, SCREENREADER]: Changes to the length of the form SHOULD be announced to screen reader users 

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#16988]
## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

<hr/>

## Point of Contact

**VFS Point of Contact:** Josh

## User Story or Problem Statement

As a screen reader user, I expect to be politely alerted of any important changes to the page so that I can plan my following steps accordingly.

## Details

Depending on the user's selections, the length of the flow may shrink or expand. Sighted users are aware of this as the heading and progress indicator on the page are updated upon selection. However, screen reader users are not notified of this change.

### Solution A: Quickly improve the screen reader experience
- [ ] The screen reader politely announces to the user when a change to the form's length has been made to provide an equal experience to sighted users.

OR 

### Solution B: Enhance the design of the experience
- [ ] An alert on the page itself announces the change and why more steps have been added.

OR

### Solution C: Simplify the experience
- [ ] Explore alternative design options.

## Environment

* Operating System: Mac, Windows
* Browser: Safari, Firefox
* Screenreading device: VoiceOver, NVDA
* Server destination: staging, production

## Steps to Recreate

1. Enter `https://staging.va.gov/careers-employment/education-and-career-counseling/apply-career-guidance-form-28-8832/status-selection` in browser.
2. Begin form without signing in.
3. Fill out the form using dummy data and reach "Step 1 of 2: Applicant Information", "Select one of the options" question.
4. Start one of the screen reading devices listed in environment.
5. Select "Spouse of a..."
6. The progress indicator expands and the `h2` string updates to "Step 1 of 3: Applicant Information."
7. Nothing is announced on the screen reader.

### Solution A: Quickly improve the screen reader experience
Simply add `aria-live: polite` to the `div` that wraps the progress indicator and `h2`. This will announce the `h2` to screen reader users if it is changed. 

``` html
<div>
     <div class="progress-bar-segmented" role="progressbar" aria-valuenow="1" aria-valuemin="0" aria-valuemax="2" tabindex="0">
       <div class="progress-segment progress-segment-complete"></div>
       <div class="progress-segment "></div>
     </div>
     <div class="schemaform-chapter-progress">
       <div aria-valuenow="1" aria-valuemin="1" aria-valuetext="Step 1 of 2: Applicant Information" aria-valuemax="2" class="nav-header nav-header-schemaform">
       <h2 id="nav-form-header" class="vads-u-font-size--h4" tabindex="-1">Step 1 of 2: Applicant Information</h2>
       </div>
     </div>
</div>
```

``` diff
<!-- Add in aria live polite to this div -->
+ <div aria-live="polite">
     <div class="progress-bar-segmented" role="progressbar" aria-valuenow="1" aria-valuemin="0" aria-valuemax="2" tabindex="0">
       <div class="progress-segment progress-segment-complete"></div>
       <div class="progress-segment "></div>
     </div>
     <div class="schemaform-chapter-progress">
       <div aria-valuenow="1" aria-valuemin="1" aria-valuetext="Step 1 of 2: Applicant Information" aria-valuemax="2" class="nav-header nav-header-schemaform">
       <h2 id="nav-form-header" class="vads-u-font-size--h4" tabindex="-1">Step 1 of 2: Applicant Information</h2>
       </div>
     </div>
 </div>
```

## Solution B: Enhance the design of the experience
Add an alert on the page that announces that more questions/steps have been added to the flow based on the selection. This solution would need guidance from @sporeboy .

## Solution C: Simplify the experience 
Consider re-exploring the current design choices in communicating progress. As of now, several accessibility and usability issues exist in regards to the progress indicator and breakdown of steps including:

- While sighted users can tell if a choice will increase the amount of choices in the form, screen reader users do not have a comparable experience as it is not announced to them.
- Some forms, like the 526 form, have a significantly unequal distribution of tasks for each step. Users may feel that a form is longer than it actually is due to the progress bar indicating progress by step as opposed to volume of questions/pages.
- The progress bar is in the tab order despite not being an interactive element. NVDA will announce the progress indicator as "progress bar 1" which provides little to no context.
- The progress indicator uses the color blue to indicate progress. Since blue is a common indicator for something being interactive, users may mistake the progress bar as a means of navigation. It also draws attention away from more important elements on the page (like the continue button and question itself).

[Gov.uk conducted a study on progress indicators](https://designnotes.blog.gov.uk/2014/07/07/do-less-problems-as-shared-spaces/) where it was observed that it was mostly ignored-- but for "those that did notice it were getting confused about what it did, and others were even feeling intimidated when seeing it.”

[Their official documentation has more details on this.](https://design-system.service.gov.uk/patterns/question-pages/#using-progress-indicators) Alternatives can include only having the amount of questions or steps listed. [Adam Silver who wrote the book Form Design Patterns also agrees with this.](https://adamsilver.io/articles/form-design-from-zero-to-hero-all-in-one-blog-post/#flow-and-order)

Usability studies could be conducted to explore alternatives such as:
- Removing the progress indicator
- Chunking out steps into smaller forms and listing number of questions instead of steps

## WCAG or Vendor Guidance (optional)

* [WCAG 2.1 4.1.3 Status Messages](https://www.w3.org/WAI/WCAG21/Understanding/status-messages.html)

## Screenshots or Trace Logs

<img width="1464" alt="Screen Shot 2020-12-07 at 10 30 30 AM" src="https://user-images.githubusercontent.com/14154792/101375109-0623a980-387d-11eb-94f5-777709e63880.png">

**Comment by jason-gcio** [comment]
`aria-live` tags should be added to the progress bars in order to achieve this but this is outside our scope to do so as we are not currently able to edit the Form System.

**Comment by joshkimux** [comment]
@jason-gcio no worries! Trevor is working on a new progress bar with the design system team which should be coming out soon 😄

**Comment by 1Copenut** [comment]
@joshkimux Take a look at this issue ticket and see if it lines up with what you're thinking: https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/357

**Comment by jason-gcio** [comment]
Follow up with Josh

**Comment by ohgreatkate** [comment]
Reviewed and task is working successfully against proposed solution A.
